accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1072,10 +1072,9 @@ If _DV_ does not include an `<interfaces>` clause then _DV_ has
 relation which was specified earlier.
```

Comment:
Here I wouldn't say it has `Object?` as superinterface. It's a supertype, but top types are supertypes of all types anyway. No need to add it, or start saying that `Object?` can be a superinterface.

Also, this mixes types and declarations (or we're not specifying what I thought we'd be specifying).

I expect an extension type to have `Object` as supertype, and to implement `Object`, if its representation type is non-nullable. For a generic extension type, that may differ between different instantiations.

An extension type *declaration* may be given `Object` as immediate superinterface if its representation type is definitely non-nullable, independent of generic instantiation.
```dart
extension type E<T>(T value) {}
```
doesn't have `Object` as an immediate declared superinterface, and the type `E<Null>` doesn't have it as a superinterface at all, but `E<int>` does. (Or that's what I thought we were specifying.)

Also notice that as written, we don't add `Object` as superinterface if there is an `implements` clause, even if that implements clause doesn't necessarily introduce `Object` as transitive superinterface.
```dart
extension type _NotObject(Object? _) {}
extension type Foo(Object _) implements _NotObject {}
```
This creates an extension type with a non-nullable representation type which doesn't implement `Object`.

That's fine with me, but let's be sure we don't think it's a problem. (It may or may not still be a subtype of `Object`, there is a distinction between "implements" and "has as supertype" which matters for **Up** calculations, which only looks at superinterface, not supertypes. Let's be absolutely clear on which property we give extension types, and when.)

Reply:
> Here I wouldn't say it has `Object?` as superinterface. It's a supertype, but top types are supertypes of all types anyway.

That's only true to the extent that we say so (subtype relationships exist because we have positively stated that they exist), and we say so for extension types in the section 'Static analysis of extension types'.

At this location we also specify that there is a superinterface relationship. This relationship is needed with `Object` in order to maintain a non-special cased treatment of extension types that are non-nullable during the Dart 1 least upper bound algorithm, when we specify which non-extension type members an extension type has, when we specify whether or not there is a conflict because there is an extension type member and a non-extension type member with the same name, when we compute the subtype relationship between `C<T1...>` and `C<S1...>`, in the definition of "a type implements a type", and probably elsewhere.

I think we need to introduce an explicit "plain" top type, say `Any`, which will allow us to eliminate several anomalies. I created https://github.com/dart-lang/language/issues/3265 in order to handle that.

Reply:
Let's return to this when we have some feedback on #3265.